### PR TITLE
TMDM-14546 [REST]: PUT /data/{containerName}/query : order_by on a Foreign Key (FK) : result not sorted

### PR DESCRIPTION
**Pull request by lpw82gf**
https://jira.talendforge.org/browse/TMDM-14546
**What is the current behavior?** (You should also link to an open issue here)
- Order by FK which reference to the Entity itself not working

**What is the new behavior?**
- Order by FK which reference to the Entity itself works normally
Following data models are covered so far:
```
EntityA
--ID
--FieldX
--FKEntityA -->EntityA.ID

EntityA
--ID_0
--ID_1
--FieldX
--FKEntityA -->EntityA.ID_0,ID_1

EntityA
--ID
--FieldX
--FieldA(ComplexType)
   --FKEntityA -->EntityA.ID

EntityA
--ID_0
--ID_1
--FieldX
--FieldA(ComplexType)
   --FKEntityA -->EntityA.ID_0,ID_1
```

**Please check if the PR fulfills these requirements**

- [x] The commit message follows Talend standard
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features) ?

**What kind of change does this PR introduce?**

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build / CI related changes
- [ ] Other... Please describe:

**Does this PR introduce a breaking change?**

- [ ] Yes
- [x] No

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...

**Other information**:

### CODE REVIEW

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StandardQueryHandler.java [hunk 1]
```diff
@@ -1772,13 +1772,20 @@ private static Criterion makeAnd(Criterion... criterions) {
...
+        } else if (isReferenceField && isContainedInMain) {
+            if (isSelfReference) {
+                addConditionForSelfReferenceField(condition, alias, ((ReferenceFieldMetadata) fieldMetadata).getReferencedField());
+            } else {
+                condition.criterionFieldNames.add(getFieldName(fieldMetadata, true));
             }
-        } else if (fieldMetadata instanceof ReferenceFieldMetadata && mainType.equals(fieldMetadata.getContainingType())) {
-            condition.criterionFieldNames.add(getFieldName(fieldMetadata, true));
+        } else if (isReferenceField && !isContainedInMain && isSelfReference) {
+            addConditionForSelfReferenceField(condition, alias, ((ReferenceFieldMetadata) fieldMetadata).getReferencedField());
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StandardQueryHandler.java:1788]: The conditions from Line 1781 to Line 1788, seems can be simplified as below:
```
} else if (isReferenceField) {
	if (isSelfReference) {
		addConditionForSelfReferenceField(condition, alias, ((ReferenceFieldMetadata) fieldMetadata).getReferencedField());
	} else if (isContainedInMain) {
		condition.criterionFieldNames.add(getFieldName(fieldMetadata, true));
	}
}
```

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StandardQueryHandler.java [hunk 2]
```diff
@@ -1789,6 +1796,28 @@ private void addCondition(FieldCondition condition, String alias, FieldMetadata
...
+
+    private void addConditionForSelfReferenceField(FieldCondition condition, String alias, FieldMetadata referencedField) {
+        if (referencedField instanceof CompoundFieldMetadata) {
+            addConditionForCompoundField(condition, alias, referencedField);
+        } else {
+            condition.criterionFieldNames.add(alias + '.' + referencedField.getName());
+        }
+    }
+
+    private boolean isSelfReference(FieldMetadata fieldMetadata) {
```
Comment (VictorLi-talendbj) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StandardQueryHandler.java:1814]: Another same name function in UserQueryBuilder.java, which is more complex to check if it is self reference, what's the difference between them?

File: org.talend.mdm.query/src/com/amalto/core/query/user/UserQueryBuilder.java [hunk 3]
```diff
@@ -929,16 +929,41 @@ public UserQueryBuilder orderBy(FieldMetadata field, OrderBy.Direction direction
...
+         * --FieldX
+         * --FieldA
+         *   --FKEntityA->EntityA
+         */
+        while(container != null && container instanceof ReferenceFieldMetadata) {
+            containingType = ((ReferenceFieldMetadata) container).getContainingType();
+            container = ((ComplexTypeMetadataImpl)containingType).getContainer();
+        }
+        return referencedType.equals(containingType);
+    }
```
Comment (hw-zhu) [inline org.talend.mdm.query/src/com/amalto/core/query/user/UserQueryBuilder.java:965]: EntiyA
         * --FieldX
         * --DerivedEntityFromFKEntityA->EntityA
For inheritance entity, Can above logic get the correct results?

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StandardQueryHandler.java [hunk 4]
```diff
@@ -1789,6 +1796,28 @@ private void addCondition(FieldCondition condition, String alias, FieldMetadata
...
+        }
+    }
+
+    private void addConditionForSelfReferenceField(FieldCondition condition, String alias, FieldMetadata referencedField) {
+        if (referencedField instanceof CompoundFieldMetadata) {
+            addConditionForCompoundField(condition, alias, referencedField);
+        } else {
+            condition.criterionFieldNames.add(alias + '.' + referencedField.getName());
+        }
+    }
```
Comment (hw-zhu) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StandardQueryHandler.java:1812]: pls make sure the `alias` is correct to each of field starting from the type, and ensures all paths to the special field are covered in time.

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StandardQueryHandler.java (hunk 1 above)
Comment (lpw82gf) [reply org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StandardQueryHandler.java:1788]: Actually they are not equal, `isReferenceField & !isContainedInMain & !isSelfReference` can't get expected result.

File: org.talend.mdm.query/src/com/amalto/core/query/user/UserQueryBuilder.java (hunk 3 above)
Comment (lpw82gf) [reply org.talend.mdm.query/src/com/amalto/core/query/user/UserQueryBuilder.java:965]: Yes

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StandardQueryHandler.java (hunk 4 above)
Comment (lpw82gf) [reply org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StandardQueryHandler.java:1812]: Yes, that has been tested, so far no regression and new cases all passed.
But there may be other cases we don't know for now.

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StandardQueryHandler.java (hunk 2 above)
Comment (lpw82gf) [reply org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StandardQueryHandler.java:1814]: Different phase, the `fieldMetadata` is different. The `StandardQueryHandler#isSelfReference` is dealing with last level Object for that field, but `UserQueryBuilder#isSelfReference` is dealing with the original Object for that field.

File: org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StandardQueryHandler.java [hunk 5]
```diff
@@ -1772,13 +1772,20 @@ private static Criterion makeAnd(Criterion... criterions) {
...
+            addConditionForCompoundField(condition, alias, fieldMetadata);
+        } else if (isReferenceField && isContainedInMain) {
+            if (isSelfReference) {
+                addConditionForSelfReferenceField(condition, alias, ((ReferenceFieldMetadata) fieldMetadata).getReferencedField());
+            } else {
+                condition.criterionFieldNames.add(getFieldName(fieldMetadata, true));
             }
-        } else if (fieldMetadata instanceof ReferenceFieldMetadata && mainType.equals(fieldMetadata.getContainingType())) {
-            condition.criterionFieldNames.add(getFieldName(fieldMetadata, true));
+        } else if (isReferenceField && !isContainedInMain && isSelfReference) {
```
Comment (htyin) [inline org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StandardQueryHandler.java:1787]: Maybe needn't add `!isContainedInMain` here.

Comment (lpw82gf) [reply org.talend.mdm.core.storage.sql/src/main/java/com/amalto/core/storage/hibernate/StandardQueryHandler.java:1787]: Seems OK

**Pull request merged**